File: C++/Sedenion Linux.rar/Sedenion/Complex Testing Console/SedenionMod.cpp
#include "Base.h"
#include "Sedenion.h"
using namespace ComplexTestingConsole;
using namespace Seden;
namespace SedenionMod
{
	inline void op(const std::wstring& str, const wchar_t* ptr, Sedenion(CMPLX_UNIT_TEST_INTERFACE* f)(const Sedenion&, double))
	{
		if (str == ptr)
		{
			Sedenion Union = Sedenion::CType_Sedenion(Base::Input(L"Union = "));
			double Value = stod(Base::Input(L"Value = "));
			Base::Output(to_wstring(std::invoke(f, Union, Value)));
		}
	};
	template <typename F = Sedenion(CMPLX_UNIT_TEST_INTERFACE*)(const Sedenion&, const Sedenion&)>
	void op(const std::wstring& str, const wchar_t* ptr, F f)
	{
		if (str == ptr)
		{
			Sedenion Union = Sedenion::CType_Sedenion(Base::Input(L"Union = "));
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			Base::Output(to_wstring(std::invoke(f, Union, Value)));
		}
	};
	template <typename F>
	void power(const std::wstring& str, const wchar_t* ptr, F f)
	{
		if (str == ptr)
		{
			Sedenion Base = Sedenion::CType_Sedenion(Base::Input(L"Base = "));
			std::int64_t Exponent = stoi64_t(Base::Input(L"Exponent = "));
			Base::Output(to_wstring(std::invoke(f, Base, Exponent)));
		}
	};
	template <typename... args>
	void power(const std::wstring& str, std::wstring&& ptr, Sedenion(CMPLX_UNIT_TEST_INTERFACE* f)(const Sedenion&, const Sedenion&, std::int64_t, args...))
	{
		if (str == ptr)
		{
			Sedenion Union = Sedenion::CType_Sedenion(Base::Input(L"Union = "));
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			std::array<std::int64_t, 1 + sizeof...(args)> Data{};
			power_get(Data);
			power_result(f, ptr, Union, Value, Data);
		}
		else if (str == ptr + L"()")
		{
			Sedenion Union = Sedenion::CType_Sedenion(Base::Input(L"Union = "));
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			std::array<std::pair<std::int64_t, std::int64_t>, 1 + sizeof...(args)> Data{};
			power_get(Data);
			power_result(f, ptr, Union, Value, Data);
		}
	};
	template <typename F>
	void basic(const std::wstring& str, const wchar_t* ptr, F f)
	{
		if (str == ptr)
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			Base::Output(to_wstring(std::invoke(f, Value)));
		}
	};
	template <typename R>
	void basic(const std::wstring& str, std::wstring&& ptr, R(CMPLX_UNIT_TEST_INTERFACE* f)(const Sedenion&, std::int64_t))
	{
		if (str == ptr)
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			std::int64_t Theta = stoi64_t(Base::Input(L"Theta = "));
			Base::Output(to_wstring(std::invoke(f, Value, Theta)));
		}
		else if (str == ptr + L"()")
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			std::int64_t ThetaMin = stoi64_t(Base::Input(L"ThetaMin = "));
			std::int64_t ThetaMax = stoi64_t(Base::Input(L"ThetaMax = "));
			for (std::int64_t Theta = ThetaMin; Theta <= ThetaMax; ++Theta)
			{
				Base::Output(ptr + L"(" + to_wstring(Theta) + L") = ", to_wstring(std::invoke(f, Value, Theta)));
			}
		}
	};
	template <typename F = Sedenion(CMPLX_UNIT_TEST_INTERFACE*)(const Sedenion&)>
	inline void tri(const std::wstring& str, const wchar_t* ptr, F f)
	{
		if (str == ptr)
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			Base::Output(to_wstring(std::invoke(f, Value)));
		}
	};
	template <typename F = Sedenion(CMPLX_UNIT_TEST_INTERFACE*)(const Sedenion&, bool, std::int64_t)>
	inline void arctri(const std::wstring& str, std::wstring&& ptr, F f)
	{
		if (str == ptr)
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			bool Sign = false;
			std::wstring Input = Base::Input(L"Sign : ");
			if (Input == L"+" && Input != L"-") { Sign = true; }
			std::int64_t Period = stoi64_t(Base::Input(L"Period = "));
			Base::Output(to_wstring(std::invoke(f, Value, Sign, Period)));
		}
		else if (str == ptr + L"()")
		{
			Sedenion Value = Sedenion::CType_Sedenion(Base::Input(L"Value = "));
			std::int64_t PeriodMin = stoi64_t(Base::Input(L"PeriodMin = "));
			std::int64_t PeriodMax = stoi64_t(Base::Input(L"PeriodMax = "));
			for (std::int64_t Period = PeriodMin; Period <= PeriodMax; ++Period)
			{
				Base::Output(ptr + L"(+, " + to_wstring(Period) + L") = ", to_wstring(std::invoke(f, Value, true, Period)));
			}
			for (std::int64_t Period = PeriodMin; Period <= PeriodMax; ++Period)
			{
				Base::Output(ptr + L"(-, " + to_wstring(Period) + L") = ", to_wstring(std::invoke(f, Value, false, Period)));
			}
		}
	};
	class MySedenionTestor final
	{
	private:
		constexpr MySedenionTestor() noexcept = delete;
		constexpr ~MySedenionTestor() noexcept = delete;
	public:
		static void Load() noexcept;
	};
	void MySedenionTestor::Load() noexcept
	{
		Base::Startup(Base::GetTitle() + L" (" + Base::GetSedenTitle() + L")");
		Base::Selection(L"=   +   -   *   /   ^   Power()   Root()   Log()");
		Base::Selection(L"abs   arg()   conjg   sgn   inverse   exp   ln()");
		Base::Selection(L"sin   cos   tan   csc   sec   cot   arcsin()   arccos()   arctan()   arccsc()   arcsec()   arccot()");
		Base::Selection(L"sinh   cosh   tanh   csch   sech   coth   arcsinh()   arccosh()   arctanh()   arccsch()   arcsech()   arccoth()");
		Base::Selection(Base::GetStartupLine());
		for (std::wstring Str; !Base::IsSwitchTo(Str); Str = Base::Input())
		{
			if (Str.empty()) { continue; }
			try
			{
				op(Str, L"=", operator ==);
				op(Str, L"+", operator +);
				op(Str, L"-", operator -);
				op(Str, L"*", operator *);
				op(Str, L"/", operator /);
				/****/
				power(Str, L"^", operator ^);
				power(Str, L"Power", Sedenion::Power);
				power(Str, L"Root", Sedenion::Root);
				power(Str, L"Log", Sedenion::Log);
				/****/
				basic(Str, L"abs", Sedenion::abs);
				basic(Str, L"arg", Sedenion::arg);
				basic(Str, L"conjg", Sedenion::conjg);
				basic(Str, L"sgn", Sedenion::sgn);
				basic(Str, L"inverse", Sedenion::inverse);
				basic(Str, L"exp", Sedenion::exp);
				basic(Str, L"ln", Sedenion::ln);
				/****/
				tri(Str, L"sin", Sedenion::sin);
				tri(Str, L"cos", Sedenion::cos);
				tri(Str, L"tan", Sedenion::tan);
				tri(Str, L"csc", Sedenion::csc);
				tri(Str, L"sec", Sedenion::sec);
				tri(Str, L"cot", Sedenion::cot);
				tri(Str, L"sinh", Sedenion::sinh);
				tri(Str, L"cosh", Sedenion::cosh);
				tri(Str, L"tanh", Sedenion::tanh);
				tri(Str, L"csch", Sedenion::csch);
				tri(Str, L"sech", Sedenion::sech);
				tri(Str, L"coth", Sedenion::coth);
				arctri(Str, L"arcsin", Sedenion::arcsin);
				arctri(Str, L"arccos", Sedenion::arccos);
				arctri(Str, L"arctan", Sedenion::arctan);
				arctri(Str, L"arccsc", Sedenion::arccsc);
				arctri(Str, L"arcsec", Sedenion::arcsec);
				arctri(Str, L"arccot", Sedenion::arccot);
				arctri(Str, L"arcsinh", Sedenion::arcsinh);
				arctri(Str, L"arccosh", Sedenion::arccosh);
				arctri(Str, L"arctanh", Sedenion::arctanh);
				arctri(Str, L"arccsch", Sedenion::arccsch);
				arctri(Str, L"arcsech", Sedenion::arcsech);
				arctri(Str, L"arccoth", Sedenion::arccoth);
			}
			catch (const std::exception& Ex) { Base::Exception(Ex); }
		}
	};
}

File: C#/Sedenion.rar/Sedenion/Complex Testing Console/SedenionMod.cs
using ComplexTestingConsole;
using Seden;
using System;
using System.Reflection;
using static ComplexTestingConsole.Module;
using static SedenionMod;
using static Seden.Sedenion;
internal static class SedenionMod
{
    internal static void Op<R>(string str, string ptr, Func<Sedenion, Sedenion, R> f)
    {
        if (str == ptr)
        {
            Sedenion Union = CType_Sedenion(Base.Input("Union = "));
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            Base.Output(f.Invoke(Union, Value).ToModuleString());
        }
    }
    internal static void Op(string str, string ptr, Func<Sedenion, double, Sedenion> f)
    {
        if (str == ptr)
        {
            Sedenion Union = CType_Sedenion(Base.Input("Union = "));
            double Value = double.Parse(Base.Input("Value = ").Replace(" ", ""));
            Base.Output(f.Invoke(Union, Value).ToModuleString());
        }
    }
    internal static void PowerOp(string str, string ptr, Func<Sedenion, long, Sedenion> f)
    {
        if (str == ptr)
        {
            Sedenion BaseNumber = CType_Sedenion(Base.Input("Base = "));
            long ExponentNumber = long.Parse(Base.Input("Exponent = ").Replace(" ", ""));
            Base.Output(f.Invoke(BaseNumber, ExponentNumber).ToModuleString());
        }
    }
    internal static void Power(string str, string ptr, Func<Sedenion, Sedenion, long, long, long, Sedenion> f)
    {
        Power(str, ptr, Delegate.CreateDelegate(f.GetType(), f.GetMethodInfo()));
    }
    internal static void Power(string str, string ptr, Func<Sedenion, Sedenion, long, long, long, long, Sedenion> f)
    {
        Power(str, ptr, Delegate.CreateDelegate(f.GetType(), f.GetMethodInfo()));
    }
    internal static void Power(string str, string ptr, Delegate f)
    {
        if (str == ptr)
        {
            Sedenion Union = CType_Sedenion(Base.Input("Union = "));
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            long[] Data = new long[f.GetMethodInfo().GetParameters().LongLength - 2];
            PowerGet(Data);
            PowerResult(f, ptr, Union, Value, Data);
        }
        else if (str == ptr + "()")
        {
            Sedenion Union = CType_Sedenion(Base.Input("Union = "));
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            Tuple<long, long>[] Data = new Tuple<long, long>[f.GetMethodInfo().GetParameters().LongLength - 2];
            PowerGet(Data);
            PowerResult(f, ptr, Union, Value, Data);
        }
    }
    internal static void Basic<R>(string str, string ptr, Func<Sedenion, R> f)
    {
        if (str == ptr)
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            Base.Output(f.Invoke(Value).ToModuleString());
        }
    }
    internal static void BasicWith<R>(string str, string ptr, Func<Sedenion, long, R> f)
    {
        if (str == ptr)
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            long Theta = long.Parse(Base.Input("Theta = ").Replace(" ", ""));
            Base.Output(f.Invoke(Value, Theta).ToModuleString());
        }
        else if (str == ptr + "()")
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            long ThetaMin = long.Parse(Base.Input("ThetaMin = ").Replace(" ", ""));
            long ThetaMax = long.Parse(Base.Input("ThetaMax = ").Replace(" ", ""));
            for (long Theta = ThetaMin; Theta <= ThetaMax; ++Theta)
            {
                Base.Output(ptr + "(" + Theta.ToModuleString() + ") = ", f.Invoke(Value, Theta).ToModuleString());
            }
        }
    }
    internal static void Tri(string str, string ptr, Func<Sedenion, Sedenion> f)
    {
        if (str == ptr)
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            Base.Output(f.Invoke(Value).ToModuleString());
        }
    }
    internal static void Arctri(string str, string ptr, Func<Sedenion, bool, long, Sedenion> f)
    {
        if (str == ptr)
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            bool Sign = false;
            string Input = Base.Input("Sign : ").Replace(" ", "");
            if (Input == "+") { Sign = true; }
            else if (Input != "-") { throw new ArgumentException("A string interpretation of the sign cannot be converted as a bool value."); }
            long Period = long.Parse(Base.Input("Period = ").Replace(" ", ""));
            Base.Output(f.Invoke(Value, Sign, Period).ToModuleString());
        }
        else if (str == ptr + "()")
        {
            Sedenion Value = CType_Sedenion(Base.Input("Value = "));
            long PeriodMin = long.Parse(Base.Input("PeriodMin = ").Replace(" ", ""));
            long PeriodMax = long.Parse(Base.Input("PeriodMax = ").Replace(" ", ""));
            for (long Period = PeriodMin; Period <= PeriodMax; ++Period)
            {
                Base.Output(ptr + "(+, " + Period.ToModuleString() + ") = ", f.Invoke(Value, true, Period).ToModuleString());
            }
            for (long Period = PeriodMin; Period <= PeriodMax; ++Period)
            {
                Base.Output(ptr + "(-, " + Period.ToModuleString() + ") = ", f.Invoke(Value, false, Period).ToModuleString());
            }
        }
    }
}
internal static class MySedenionTestor
{
    internal static void Load()
    {
        Base.Startup(Base.GetTitle() + " (" + Base.GetSedenTitle() + ")");
        Base.Selection("=   +   -   *   /   ^   Power()   Root()   Log()");
        Base.Selection("abs   arg()   conjg   sgn   inverse   exp   ln()");
        Base.Selection("sin   cos   tan   csc   sec   cot   arcsin()   arccos()   arctan()   arccsc()   arcsec()   arccot()");
        Base.Selection("sinh   cosh   tanh   csch   sech   coth   arcsinh()   arccosh()   arctanh()   arccsch()   arcsech()   arccoth()");
        Base.Selection(Base.GetStartupLine());
        for (string Str = ""; !Base.IsSwitchTo(Str); Str = Base.Input())
        {
            if (Str == "") { continue; }
            try
            {
                Op(Str, "=", (Sedenion Union, Sedenion Value) => { return Union == Value; });
                Op(Str, "+", (Sedenion Union, Sedenion Value) => { return Union + Value; });
                Op(Str, "-", (Sedenion Union, Sedenion Value) => { return Union - Value; });
                Op(Str, "*", (Sedenion Union, Sedenion Value) => { return Union * Value; });
                Op(Str, "/", (Sedenion Union, double Value) => { return Union / Value; });
                /****/
                PowerOp(Str, "^", (Sedenion Union, long Value) => { return Union ^ Value; });
                Power(Str, "Power", Power);
                Power(Str, "Root", Root);
                Power(Str, "Log", Log);
                /****/
                Basic(Str, "abs", Abs);
                BasicWith(Str, "arg", Arg);
                Basic(Str, "conjg", Conjg);
                Basic(Str, "sgn", Sgn);
                Basic(Str, "inverse", Inverse);
                Basic(Str, "exp", Exp);
                BasicWith(Str, "ln", Ln);
                /****/
                Tri(Str, "sin", Sin);
                Tri(Str, "cos", Cos);
                Tri(Str, "tan", Tan);
                Tri(Str, "csc", Csc);
                Tri(Str, "sec", Sec);
                Tri(Str, "cot", Cot);
                Tri(Str, "sinh", Sinh);
                Tri(Str, "cosh", Cosh);
                Tri(Str, "tanh", Tanh);
                Tri(Str, "csch", Csch);
                Tri(Str, "sech", Sech);
                Tri(Str, "coth", Coth);
                Arctri(Str, "arcsin", Arcsin);
                Arctri(Str, "arccos", Arccos);
                Arctri(Str, "arctan", Arctan);
                Arctri(Str, "arccsc", Arccsc);
                Arctri(Str, "arcsec", Arcsec);
                Arctri(Str, "arccot", Arccot);
                Arctri(Str, "arcsinh", Arcsinh);
                Arctri(Str, "arccosh", Arccosh);
                Arctri(Str, "arctanh", Arctanh);
                Arctri(Str, "arccsch", Arccsch);
                Arctri(Str, "arcsech", Arcsech);
                Arctri(Str, "arccoth", Arccoth);
            }
            catch (Exception Ex) { Base.Exception(Ex); }
        }
    }
}
